### PR TITLE
Adds DebertaV2/V3

### PR DESCRIPTION
**Pull request by BradyBonnette**
This adds Microsoft's Deberta V2 and V3 model into the `candle` ecosystem. It also includes an example file demonstrating how to use it, as well as a README for more information.

At the time of this commit, the model can only do Named Entity Recognition and Text Classification. There are other modes such as Question Answering, Multiple Choice and Masked Input that could still be developed for this at a later point in time.

### CODE REVIEW

File: candle-examples/examples/debertav2/main.rs [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+                });
+            }
+
+            println!("\n{:?}", results);
+        }
+    }
+    Ok(())
+}
+
+fn create_benchmark<F>(
```
Comment (BradyBonnette) [inline candle-examples/examples/debertav2/main.rs:369]: I saw that `criterion` is in the workspace Cargo.toml file as a dependency, but I did not see it in the `candle-examples` crate. I wasn't sure if the inclusion of criterion in the examples was warranted or wanted.

File: candle-transformers/src/models/debertav2.rs [hunk 2]
```diff
@@ -0,0 +1,1499 @@
...
+enum PositionEmbeddingType {
+    #[default]
+    Absolute,
+}
+
+pub type Id2Label = HashMap<u32, String>;
+pub type Label2Id = HashMap<String, u32>;
+
+#[derive(Debug, Clone, PartialEq, Deserialize)]
+pub struct Config {
```
Comment (BradyBonnette) [inline candle-transformers/src/models/debertav2.rs:53]: This looks a lot like the normal BERT configuration, but in going through the Python Transformers code I realized there were a lot of other configuration tidbits that didn't exist in other BERT models. So, I just started over with a new one just for Deberta.

**Comment by BradyBonnette** [comment]
Using a simple Python example, such as:

```python
from transformers import pipeline
pipe = pipeline("token-classification", model="Clinical-AI-Apollo/Medical-NER")
result = pipe('45 year old woman diagnosed with CAD')
print(f"{result}")
```

This produces the following results:

```
[{'entity': 'B-AGE', 'score': 0.44746688, 'index': 1, 'word': '▁45', 'start': 0, 'end': 2},
 {'entity': 'I-AGE', 'score': 0.5699233, 'index': 2, 'word': '▁year', 'start': 2, 'end': 7},
 {'entity': 'I-AGE', 'score': 0.6126752, 'index': 3, 'word': '▁old', 'start': 7, 'end': 11},
 {'entity': 'B-SEX', 'score': 0.40775478, 'index': 4, 'word': '▁woman', 'start': 11, 'end': 17},
 {'entity': 'B-DISEASE_DISORDER', 'score': 0.34644428, 'index': 7, 'word': '▁CAD', 'start': 32, 'end': 36}]
```

This is in comparison with using the `example.rs` file provided by this PR:

```bash
cargo run  
  --example debertav2 
  --release 
  --features=cuda,cudnn -- 
  --model-id=Clinical-AI-Apollo/Medical-NER 
  --revision=main 
  --sentence='45 year old woman diagnosed with CAD'
```

results in:

```
[[NERItem { entity: "B-AGE", word: "▁45", score: 0.44766188, start: 0, end: 2, index: 1 },
 NERItem { entity: "I-AGE", word: "▁year", score: 0.5694605, start: 2, end: 7, index: 2 },
 NERItem { entity: "I-AGE", word: "▁old", score: 0.6122175, start: 7, end: 11, index: 3 },
 NERItem { entity: "B-SEX", word: "▁woman", score: 0.40786597, start: 11, end: 17, index: 4 },
 NERItem { entity: "B-DISEASE_DISORDER", word: "▁CAD", score: 0.34624502, start: 32, end: 36, index: 7 }]]
```

There's a tiny amount of precision difference between the Python and Rust versions, but from my understanding it's so insignficant that it does not make a difference in accuracy nor performance.

File: candle-examples/examples/debertav2/main.rs (hunk 1 above)
Comment (LaurentMazare) [reply candle-examples/examples/debertav2/main.rs:369]: Current version is fine, we prefer not having too much dependencies on the examples so avoid using criterion there.

Comment (BradyBonnette) [reply candle-examples/examples/debertav2/main.rs:369]: That's what I was assuming. I don't even know what really compelled me to have a benchmark in there, but I figured it'd be good to have to compare against any Python version.

File: candle-examples/examples/debertav2/README.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+Note that all examples here use the `cuda` and `cudnn` feature flags provided by the `candle-examples` crate. You may need to adjust them to match your environment.
+
+### NER / Token Classification
+
+NER is the default task provided by this example if the `--task` flag is not set.
+
+To use a model from HuggingFace hub (as seen at https://huggingface.co/blaze999/Medical-NER):
+
+```bash
+cargo run  --example debertav2 --release --features=cuda,cudnn -- --model-id=blaze999/Medical-NER --revision=main --sentence='63 year old woman with history of CAD presented to ER'
```
Comment (LaurentMazare) [inline candle-examples/examples/debertav2/README.md:16]: cudnn is useless for such a text model, so need to have it as a feature here.

File: candle-examples/examples/debertav2/main.rs [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+        let config: DebertaV2Config = serde_json::from_str(&config)?;
+
+        // Command-line id2label takes precedence. Otherwise, use model config's id2label.
+        // If neither is specified, then we can't proceed.
+        let id2label = if let Some(id2labelstr) = &self.id2label {
+            serde_json::from_str(id2labelstr.as_str())?
+        } else if let Some(id2label) = &config.id2label {
+            id2label.clone()
+        } else {
+            return Err(Error::msg(
```
Comment (LaurentMazare) [inline candle-examples/examples/debertav2/main.rs:149]: Slightly simpler to use `anyhow::bail!` to generate the error.

File: candle-transformers/src/models/debertav2.rs [hunk 5]
```diff
@@ -0,0 +1,1499 @@
...
+    embedding_size: usize,
+    embed_proj: Option<candle_nn::Linear>,
+}
+
+impl DebertaV2Embeddings {
+    pub fn load(vb: VarBuilder, config: &Config) -> candle::Result<Self> {
+        let device = vb.device().clone();
+        let config = config.clone();
+
+        let embedding_size = match config.embedding_size {
```
Comment (LaurentMazare) [inline candle-transformers/src/models/debertav2.rs:144]: `unwrap_or`

File: candle-transformers/src/models/debertav2.rs [hunk 6]
```diff
@@ -0,0 +1,1499 @@
...
+        let position_embeddings = match config.position_biased_input {
+            true => Some(embedding(
+                config.max_position_embeddings,
+                embedding_size,
+                vb.pp("position_embeddings"),
+            )?),
+            false => None,
+        };
+
+        let token_type_embeddings: Option<Embedding> = match config.type_vocab_size > 0 {
```
Comment (LaurentMazare) [inline candle-transformers/src/models/debertav2.rs:161]: Use `if` rather than `match` for such conditions. The same applies below.

File: candle-transformers/src/models/debertav2.rs [hunk 7]
```diff
@@ -0,0 +1,1499 @@
...
+        token_type_ids: Option<&Tensor>,
+        position_ids: Option<&Tensor>,
+        mask: Option<&Tensor>,
+        inputs_embeds: Option<&Tensor>,
+    ) -> candle::Result<Tensor> {
+        let input_shape = match (input_ids, inputs_embeds) {
+            (Some(inputids), None) => inputids.dims(),
+            (None, Some(inputsembeds)) => inputsembeds.dims(),
+            (None, None) => {
+                return Err(candle::Error::Msg(
```
Comment (LaurentMazare) [inline candle-transformers/src/models/debertav2.rs:216]: `bail!`

File: candle-transformers/src/models/debertav2.rs [hunk 8]
```diff
@@ -0,0 +1,1499 @@
...
+
+        let query_layer = self.transpose_for_scores(&self.query_proj.forward(query_states)?)?;
+        let key_layer = self.transpose_for_scores(&self.key_proj.forward(query_states)?)?;
+        let value_layer = self.transpose_for_scores(&self.value_proj.forward(query_states)?)?;
+
+        let mut rel_att: Option<Tensor> = None;
+
+        let mut scale_factor: usize = 1;
+
+        if self.config.pos_att_type.contains(&"c2p".to_string()) {
```
Comment (LaurentMazare) [inline candle-transformers/src/models/debertav2.rs:435]: Wouldn't it work to just use `"c2p"` rather than making a string and referencing it?

File: candle-transformers/src/models/debertav2.rs [hunk 9]
```diff
@@ -0,0 +1,1499 @@
...
+        let dims = xs.dims().to_vec();
+        let result = match dims.len() {
+            3 => {
+                let reshaped = xs.reshape((dims[0], dims[1], self.num_attention_heads, ()))?;
+
+                let new_dims = reshaped.dims();
+
+                reshaped.transpose(1, 2)?.contiguous()?.reshape((
+                    (),
+                    new_dims[1],
```
Comment (LaurentMazare) [inline candle-transformers/src/models/debertav2.rs:541]: `new_dims.dim(1)?` make a proper error rather than a panic if there is an issue.

File: candle-transformers/src/models/debertav2.rs [hunk 10]
```diff
@@ -0,0 +1,1499 @@
...
+                    *new_dims.last().unwrap(),
```
Comment (LaurentMazare) [inline candle-transformers/src/models/debertav2.rs:542]: `new_dims.dim(D::Minus1)?`

File: candle-transformers/src/models/debertav2.rs [hunk 11]
```diff
@@ -0,0 +1,1499 @@
...
+                ))
+            }
+            shape => Err(candle::Error::Msg(format!(
```
Comment (LaurentMazare) [inline candle-transformers/src/models/debertav2.rs:545]: `bail!`

File: candle-transformers/src/models/debertav2.rs [hunk 12]
```diff
@@ -0,0 +1,1499 @@
...
+            )?
+        } else {
+            relative_pos.cloned().unwrap()
+        };
+
+        relative_pos = match relative_pos.dims().len() {
+            2 => relative_pos.unsqueeze(0)?.unsqueeze(0)?,
+            3 => relative_pos.unsqueeze(1)?,
+            other => {
+                return Err(candle::Error::Msg(format!(
```
Comment (LaurentMazare) [inline candle-transformers/src/models/debertav2.rs:578]: `bail!`

File: candle-transformers/src/models/debertav2.rs [hunk 13]
```diff
@@ -0,0 +1,1499 @@
...
+                self.transpose_for_scores(&self.query_proj.forward(&rel_embeddings)?)?
+                    .repeat(repeat_with)?,
+            );
+
+            pos_key_layer = Some(
+                self.transpose_for_scores(&self.key_proj.forward(&rel_embeddings)?)?
+                    .repeat(repeat_with)?,
+            )
+        } else {
+            if self.config.pos_att_type.contains(&"c2p".to_string()) {
```
Comment (LaurentMazare) [inline candle-transformers/src/models/debertav2.rs:605]: Same as above.

File: candle-transformers/src/models/debertav2.rs [hunk 14]
```diff
@@ -0,0 +1,1499 @@
...
+                        .expand(&[
+                            query_layer.dim(0)?,
+                            key_layer.dim(D::Minus2)?,
+                            key_layer.dim(D::Minus2)?,
+                        ])?
+                        .contiguous()?
+                        .to_dtype(DType::U32)?,
+                    D::Minus1,
+                )?
+                .transpose(D::Minus1, D::Minus2)?;
```
Comment (LaurentMazare) [inline candle-transformers/src/models/debertav2.rs:715]: That's just `.t()?`.

File: candle-transformers/src/models/debertav2.rs [hunk 15]
```diff
@@ -0,0 +1,1499 @@
...
+            output_states = layer_module.forward(
+                next_kv.as_ref(),
+                &attention_mask,
+                query_states.as_ref(),
+                relative_pos.as_ref(),
+                rel_embeddings.as_ref(),
+            )?;
+
+            if i == 0 && self.conv.is_some() {
+                output_states = self.conv.as_ref().unwrap().forward(
```
Comment (LaurentMazare) [inline candle-transformers/src/models/debertav2.rs:1089]: Do a pattern matching on `self.conv.as_ref` rather than `is_some` then `unwrap`.

**Review by LaurentMazare — COMMENTED** [review]
Looks pretty good. I've put some mostly cosmetic comments in line, would be great if you can apply them to the whole file as I didn't bother repeating them. Mostly we should avoid everything that can panic, i.e. make actual errors rather than `unwrap`, use `bail!` to shorten error generation, etc.

**Comment by BradyBonnette** [comment]
> Looks pretty good. I've put some mostly cosmetic comments in line, would be great if you can apply them to the whole file as I didn't bother repeating them. Mostly we should avoid everything that can panic, i.e. make actual errors rather than `unwrap`, use `bail!` to shorten error generation, etc.

No problem, I can do that later today. Thanks for reviewing it, and if you see any other things that should change please let me know.

File: candle-transformers/src/models/debertav2.rs (hunk 8 above)
Comment (BradyBonnette) [reply candle-transformers/src/models/debertav2.rs]: The way I have it set up in the config is that `pos_att_type` is parsed into a `Vec<String>` by basically splitting the string on a pipe character. Or, if it's specified as a JSON array in the config, it will be translated directly into a `Vec<String>`. I addressed this in an update to use an iterator and check for `any()`.

File: candle-transformers/src/models/debertav2.rs (hunk 13 above)
Comment (BradyBonnette) [reply candle-transformers/src/models/debertav2.rs]: See above

**Review by LaurentMazare — APPROVED** [review]

**Comment by LaurentMazare** [comment]
Thanks for the PR! I've made some small tweaks to avoid some cases with `foo.unwrap_or(bar)` as in this case the value `bar` is generated even if `foo` is `Some(_)` (`unwrap_or_else` gets around this), in particular there were some `bar` expressions that contained some `unwrap` that could have likely failed.

**Comment by Stealthwriter** [comment]
Deberta v3 large is the best model out there for text classification, it's top rated in kaggle competitions
